src/Orleans/Messaging/ProxiedMessageCenter.cs

```diff
@@ -287,6 +293,12 @@ public Message WaitMessage(Message.Categories type, CancellationToken ct)
         {
             try
             {
+                if (ct.IsCancellationRequested)
+                {
+                    return null;
+                }
+
+                // Don't pass CancellationToken to Take. It causes too much spinning.
                 Message msg = PendingInboundMessages.Take();
```

Comment:
Sure, although this actually might have some unmanaged resources, since it's using semaphores. Still it might be fine.
Catching  InvalidOperationException still makes sense though according to docs